### PR TITLE
add segmentation based (dice) loss

### PR DESCRIPTION
**Pull request by aymuos15**
As proposed in https://github.com/google-deepmind/optax/issues/1358, it would be nice to have segmentation based losses as well. A quick google search (and perplexity's search) shows that there really isn't a standard implementation for dice loss in JAX.

The issue mentions DSC + CE as well. However, I am actually not very sure how exactly this wants to be implemented, given that there is no great reference standard to go by. My personal suggestion would be to just focus on dice first, and then do a separate PR for a follow up of including more variants. So I would love some advice on what would be the best way to proceed with this. Happy to include/remove anything.

For a first pass, I tried to cover the basic bases which I generally do, and keep a reasonable amount of tests to cover said bases.

Just to note: If it means anything to anyone here, there really isn't a standard followed implementation part of the torch ecosystem either.

### CODE REVIEW

**Comment by aymuos15** [comment]
Hi @rdyro . This is my first pass. I'll fix the lint issue (wasn't aware of this specific one, will keep in mind) based on what we decide here if thats okay. Thanks again :)

**Comment by vroulet** [comment]
Thank you very much @aymuos15 ! Looks pretty good to me.

@selamw1 or @rajasekharporeddy would you have time to thoroughly review this PR? Thanks !

File: docs/api/losses.rst [hunk 1]
```diff
@@ -32,6 +32,10 @@ Losses
     sparsemax_loss
     squared_error
     triplet_margin_loss
+    dice_loss
```
Comment (rajasekharporeddy) [inline docs/api/losses.rst]: Keep functions in alphabetical order

File: docs/api/losses.rst [hunk 2]
```diff
@@ -120,3 +124,9 @@ Sparsemax
 Triplet margin loss
 ~~~~~~~~~~~~~~~~~~~
 .. autofunction:: triplet_margin_loss
+
```
Comment (rajasekharporeddy) [inline docs/api/losses.rst]: Keep functions in alphabetical order

File: optax/losses/_segmentation.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+) -> chex.Array:
+    """Computes the Dice Loss for multi-class segmentation.
+
+    Computes the Soft Dice Loss for segmentation tasks. Works for both binary
+    and multi-class segmentation. For binary segmentation, use targets with
+    shape [..., 1] or [...] and predictions with corresponding logits.
+
+    The loss is computed per class and then averaged (or summed) across classes.
+    For class c: dice_c = (2 * intersection_c + smooth) /
+                         (pred_c + target_c + smooth)
```
Comment (rajasekharporeddy) [inline optax/losses/_segmentation.py]:
```suggestion
    For class ``c``: 

    .. math::
      dice_c = (2 * intersection_c + smooth) / (pred_c + target_c + smooth)
```

File: optax/losses/_segmentation.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+def dice_loss(
+    predictions: chex.Array,
+    targets: chex.Array,
+    *,
+    class_weights: Optional[chex.Array] = None,
+    smooth: float = 1.0,
+    apply_softmax: bool = True,
+    reduction: str = "mean",
+    ignore_background: bool = False,
+) -> chex.Array:
```
Comment (rajasekharporeddy) [inline optax/losses/_segmentation.py:34]: I believe it would be more robust to provide `axis` as an explicit argument, allowing `spatial_axes` to be dynamically determined within the function. WDYT?

File: optax/losses/_segmentation.py (hunk 3 above)
Comment (rajasekharporeddy) [reply optax/losses/_segmentation.py]: Also, define `intersection_c` here.

File: optax/losses/_segmentation.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    # Convert logits to probabilities
+    if apply_softmax:
+        if predictions.shape[-1] == 1:
+            # Binary case - use sigmoid
+            probs = jax.nn.sigmoid(predictions)
+        else:
+            # Multi-class case - use softmax
+            probs = jax.nn.softmax(predictions, axis=-1)
+    else:
+        probs = predictions
```
Comment (rajasekharporeddy) [inline optax/losses/_segmentation.py]: Use `jax.lax.cond` to remove these conditional branches and pass the `pylint` tests.
```suggestion
    probs = jax.lax.cond(
        apply_softmax,
        lambda preds: jax.lax.cond(
            preds.shape[-1] == 1,
            jax.nn.sigmoid,
            lambda p: jax.nn.softmax(p, axis=-1),
            preds
        ),
        lambda preds: preds,
        predictions
    )
```

File: optax/losses/_segmentation.py (hunk 4 above)
Comment (aymuos15) [reply optax/losses/_segmentation.py:34]: I do agree with this. Will update this in the next commit. Thanks for pointing that out.

**Comment by aymuos15** [comment]
The doctests were failing hence the repeated commits.

Thank you very much @rajasekharporeddy for the review.

**Comment by rajasekharporeddy** [comment]
@vroulet The PR looks good to me.

File: optax/losses/_segmentation.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+                f"Found sum range: [{jnp.min(pred_sums):.6f}, "
+                f"{jnp.max(pred_sums):.6f}]"
+            )
+
+    # Convert logits to probabilities using jax.lax.cond to avoid control flow
+    probs = jax.lax.cond(
+        apply_softmax,
+        # True branch: apply sigmoid or softmax based on number of classes
+        lambda: jax.lax.cond(
+            predictions.shape[-1] == 1,
```
Comment (rdyro) [inline optax/losses/_segmentation.py]: This should probably be python control flow since it's a static condition.

File: optax/losses/_segmentation.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+            raise ValueError(
+                "When apply_softmax=False, predictions must be valid "
+                "probability distributions that sum to 1 along the class axis. "
+                f"Found sum range: [{jnp.min(pred_sums):.6f}, "
+                f"{jnp.max(pred_sums):.6f}]"
+            )
+
+    # Convert logits to probabilities using jax.lax.cond to avoid control flow
+    probs = jax.lax.cond(
+        apply_softmax,
```
Comment (rdyro) [inline optax/losses/_segmentation.py]: Likewise static condition, python control flow should be enough

File: optax/losses/_segmentation.py (hunk 6 above)
Comment (aymuos15) [reply optax/losses/_segmentation.py]: Ah okay. The pylint test was failing here and that's why this was done. Will try to revert this in a way the tests don't fail. Thanks!

Comment (rdyro) [reply optax/losses/_segmentation.py]: Please revert. If pylint still complains about too many conditionals you can try refactoring a little (it just counts if statements and their nesting within a single parent scope).

If it's too annoying still, feel free to add a pylint ignore comment [https://pylint.pycqa.org/en/latest/user_guide/messages/message_control.html](https://pylint.pycqa.org/en/latest/user_guide/messages/message_control.html)

Comment (aymuos15) [reply optax/losses/_segmentation.py]: Thank you very much for the info and the suggestions.

File: optax/losses/_segmentation.py [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+
+from typing import Optional
+
+import chex
+import jax
+import jax.numpy as jnp
+
+
+def dice_loss(
+    predictions: chex.Array,
```
Comment (Logon27) [inline optax/losses/_segmentation.py:25]: If these are the logits I think you should actually call the input `logits`. A lot of other optax functions have the parameter named like that. Just my opinion. I am gonna try out the code from this PR for my own uses and see how it goes.

Comment (aymuos15) [reply optax/losses/_segmentation.py:25]: That's fair. Thanks! Please update here if you face any issues. Then I will accordingly update this. Otherwise I will just make that small change.

Comment (Logon27) [reply optax/losses/_segmentation.py:25]: Oh I see you have an apply_softmax input parameter. Is there a scenario where you wouldn't want softmax applied? Just curious.

Also I tried out the binary dice loss yesterday and it seems to be working fine.

Comment (aymuos15) [reply optax/losses/_segmentation.py:25]: So I had a re-look at the code (sorry Its been kinda long since this started haha). 

Given that the the apply_softmax exists, I think ill resort to no change at this point.

There are many non-standard scenarios where you would need the logits themselves and hence i guess it is a good practice?

**Comment by rdyro** [comment]
@aymuos15 could you reindent your changes to use indent-width=2?

This is needed to pass the internal checks, but I want you to get proper credit in git history.

**Comment by Logon27** [comment]
Also it does not have to be this PR. But I think we should add tversky loss as well. Its basically dice loss but you can weight false positives and false negatives using input parameters.

File: optax/losses/_segmentation.py (hunk 8 above)
Comment (Logon27) [reply optax/losses/_segmentation.py:25]: Thats fair. I was just wondering what kind of scenarios that would be. No need for changes.

**Comment by aymuos15** [comment]
> @aymuos15 could you reindent your changes to use indent-width=2?
> 
> This is needed to pass the internal checks, but I want you to get proper credit in git history.

@rdyro Really appreciate the proper credit, thank you very much.

However, apologies, but this maybe an extremely naive question -- Could you please let me know a way to check for this? Looking around, I could not find a way to confirm through a linter about the index-width. I found yapf and autopep8 but they show that the other files (within the repo) are not completely using indent-width=2 so im just a bit confused.

@Logon27 I was planning to do another very small PR just for Dice + CE (and I think that should be more than enough because these two are the most common losses) anyways. I will include this as well. Thank you!

**Comment by rdyro** [comment]
> However, apologies, but this maybe an extremely naive question -- Could you please let me know a way to check for this? Looking around, I could not find a way to confirm through a linter about the index-width. I found yapf and autopep8 but they show that the other files (within the repo) are not completely using indent-width=2 so im just a bit confused.

I unfortunately don't know how to set up a linter for this. You can try formatting your code with pyink (a google-compatible code formatter based on black):
```bash
$ pip install pyink
$ pyink --pyink --pyink-indentation=2 --pyink-use-majority-quotes --line-length=80 {FILENAME}
```
@aymuos15

**Comment by aymuos15** [comment]
This is what I get with the current state of the commit

```bash
pyink --check --pyink-indentation=2 --pyink-use-majority-quotes --line-length=80 _segmentation.py 
pyink --check --pyink-indentation=2 --pyink-use-majority-quotes --line-length=80 _segmentation_test.py
pyink --check --pyink-indentation=2 --pyink-use-majority-quotes --line-length=80 __init__.py
All done! ✨ 🍰 ✨
1 file would be left unchanged.
All done! ✨ 🍰 ✨
1 file would be left unchanged.
All done! ✨ 🍰 ✨
1 file would be left unchanged.
```
